rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+            self.water_occupancy = 0.0;
+            self.material = Material::Water;
+            return;
+        }
+
+        self.solid_occupancy = solid_occupancy;
+
+        // Full with a solid (non-air) material? We can't have any water.
+        if solid_occupancy == 1.0 && water_occupancy > 0.0 {
+            self.water_occupancy = 0.0
+        } else {
+            self.water_occupancy = water_occupancy;
+        }
+    }
+
+    pub fn set_material(&mut self, material: Material) {
+        self.material = material;
+
+        // Occupancy determination depends on material.
+        self.set_occupancy(self.solid_occupancy, self.water_occupancy)
+    }
+}
+
+// We don't iterate over the grid, so we can use a HashMap instead of BTreeMap to save performance.
+/// A container for a chunk of terrain, used in the `Terrain` object.
+#[derive(Debug, Default, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Chunk {
+    grid: HashMap<VoxelCoordinates, Voxel>,
+    /// For all empty voxels in the chunk, we will write this material
+    /// at 100% occupancy. Defaults to `Material::Air`.
+    pub base_material: Material,
+}
+
+impl Chunk {
+    /// Constructs a new `Chunk` with a base material of `Material::Air`.
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            grid: HashMap::new(),
+            base_material: Material::Air,
+        }
+    }
+
+    /// Constructs a new `Chunk` with a user-provided base material.
+    #[inline]
+    pub fn new_with_base(base_material: Material) -> Self {
+        Self {
+            grid: HashMap::new(),
+            base_material,
+        }
+    }
+
+    /// Finds a `Voxel` at the given position in this `Chunk`,
+    /// returning a reference to it if it exists.
+    #[inline]
+    pub fn get_voxel(&self, position: &VoxelCoordinates) -> Option<&Voxel> {
+        self.grid.get(position)
+    }
+
+    /// Finds a `Voxel` at the given position in this `Chunk`,
+    /// returning a mutable reference to it if it exists.
+    #[inline]
+    pub fn get_voxel_mut(&mut self, position: &VoxelCoordinates) -> Option<&mut Voxel> {
+        self.grid.get_mut(position)
+    }
+
+    /// Writes (or overwrites) a `Chunk` at the given position to this `Terrain`.
+    #[inline]
+    pub fn write_voxel(&mut self, position: &VoxelCoordinates, voxel: Voxel) {
+        self.grid.insert(*position, voxel);
+    }
+}
+
+impl TerrainSerializer for Chunk {
+    fn encode(&self) -> Vec<u8> {
+        // ~256 bytes if all voxels are air/base mat with maximum count. Double it
+        let mut data = Vec::<u8>::with_capacity(512);
+
+        let base_voxel: Voxel = Voxel {
+            solid_occupancy: 1.0,
+            water_occupancy: 0.0,
+            material: self.base_material,
+        };
+
+        let mut pos_cursor = VoxelCoordinates::default();
+        let mut run_length_cursor = (0u16, &base_voxel);
```

Comment:
Storing these values in a tuple isn't a great idea. I would rather we just have two different variables for them.